jupyter_collaboration/rooms.py

```diff
@@ -212,49 +219,60 @@ def _on_document_change(self, target: str, event: Any) -> None:
-        )
+        asyncio.create_task(self._maybe_save_document())
 
-    async def _maybe_save_document(self, saving_document: asyncio.Task | None) -> None:
+    async def _maybe_save_document(self) -> None:
         """
         Saves the content of the document to disk.
 
         ### Note:
-            There is a save delay to debounce the save since we could receive a high
+            There is a save delay to throttle the save since we could receive a high
             amount of changes in a short period of time. This way we can cancel the
             previous save.
         """
         if self._save_delay is None:
+            # TODO: fix this; if _save_delay is unset, then this never writes to disk
             return
 
-        if saving_document is not None and not saving_document.done():
-            # the document is being saved, cancel that
-            saving_document.cancel()
+        if self._waiting_to_save:
+            # if a previously spawned `self._maybe_save_document()` task is
+            # waiting to save, then that task will save the Ydoc within
+            # `self._save_delay` seconds. therefore we can return early.
+            return
+        
+        if self._save_task and not self._save_task.done():
+            # if we are currently saving, then set the `_should_resave`
+            # flag. this indicates to the currently running `self._save_task`
+            # that it should re-call this method after it completes.
+            self._should_resave = True
+            return
+        
+        # save after `self._save_delay` seconds of inactivity
+        self._waiting_to_save = True
+        await asyncio.sleep(self._save_delay)
+        self._waiting_to_save = False
 
         # all async code (i.e. await statements) must be part of this try/except block
         # because this coroutine is run in a cancellable task and cancellation is handled here
-
         try:
-            # save after X seconds of inactivity
-            await asyncio.sleep(self._save_delay)
-
+            # do not write to `self._document` in this `try` block, as that will
+            # trigger the observer and result in a save loop.
             self.log.info("Saving the content from room %s", self._room_id)
-            await self._file.maybe_save_content(
+            loop = asyncio.get_running_loop()
+            self._save_task = loop.create_task(self._file.maybe_save_content(
                 {
                     "format": self._file_format,
                     "type": self._file_type,
                     "content": self._document.source,
                 }
-            )
-            async with self._update_lock:
-                self._document.dirty = False
-
+            ))
+            await self._save_task
             self._emit(LogLevel.INFO, "save", "Content saved.")
 
+            if self._should_resave:
+                self._should_resave = False
+                asyncio.create_task(self._maybe_save_document())
```

Comment:
Always [save a reference to the task](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task).

Reply:
Fixed in latest revision.